mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs

```diff
@@ -0,0 +1,727 @@
+    }
+
+    fn batch(&self, count: usize) -> StorageResult<VecDeque<(Key, Publication)>> {
+        let write_index = self.pointers.write;
+        let read_index = self.pointers.read;
+        let block_size = serialized_block_size()?;
+
+        // If read would go into where writes are happening then we don't have data to read.
+        if read_index == write_index {
+            return Ok(VecDeque::new());
+        }
+
+        let mut start = read_index;
+        let mut vdata = VecDeque::new();
+        for _ in 0..count {
+            // This case shouldn't be pending, as we must have gotten something we can process.
+            if start >= write_index {
+                break;
+            }
+
+            let end = start + block_size;
+            if end > self.max_file_size {
+                return Err(RingBufferError::WrapAround.into());
+            }
+
+            let bytes = &self.mmap[start..end];
+            // this is unused memory
+            if bytes.iter().map(|&x| x as usize).sum::<usize>() == 0 {
+                break;
+            }
+
+            let block = binary_deserialize::<BlockHeaderWithHash>(bytes)?;
+            // this means we read bytes that don't make a block, this is
+            // a really bad state to be in as somehow the pointers don't
+            // match to where data really is at.
+            if block.inner().hint() != BLOCK_HINT {
+                return Err(RingBufferError::Validate(BlockError::Hint).into());
+            }
+
+            let inner_block = block.inner();
+            let data_size = inner_block.data_size();
+            let index = inner_block.write_index();
+            start += block_size;
+            let end = start + data_size;
+            let data = load_data(self.mmap, start, data_size, self.max_file_size)?;
+            start = end % self.max_file_size;
+
+            // to move to beginning of next block header
+            // start += 1;
+            validate(&block, &data)?;
+            let publication = binary_deserialize::<Publication>(data.inner())?;
+            let key = Key {
+                offset: index as u64,
+            };
+            vdata.push_back((key, publication));
+        }
+
+        Ok(vdata)
+    }
+
+    fn remove(&mut self, key: usize) -> StorageResult<()> {
+        let timer = Instant::now();
+        let read_index = self.pointers.read;
+        if key != read_index {
+            return Err(StorageError::RingBuffer(RingBufferError::RemovalIndex));
+        }
+        let block_size = serialized_block_size()?;
+
+        let start = key;
+        let mut end = start + block_size;
+        if end > self.max_file_size {
+            return Err(StorageError::RingBuffer(RingBufferError::WrapAround));
+        }
+
+        let maybe_block = load_block_header(self.mmap, start, block_size, self.max_file_size);
+        if maybe_block.is_err() {
+            return Err(StorageError::RingBuffer(RingBufferError::NonExistantKey));
```

Comment:
nit: I'm just trying to avoid `.unwrap()` as much as possible.
```rust
        let mut block = match load_block_header(self.mmap, start, block_size, self.max_file_size) {
            Ok(block) => block,
            Err(e) => {
                return Err(StorageError::Serialization(e));
            }
        };

        if block.inner().hint() != BLOCK_HINT {
            return Err(StorageError::RingBuffer(RingBufferError::NonExistantKey));
        }
```

Reply:
or with `map_err`
```rust
        let mut block = match load_block_header(self.mmap, start, block_size, self.max_file_size)
           .map_err(|e| Err(StorageError::Serialization(e))?;

        if block.inner().hint() != BLOCK_HINT {
            return Err(StorageError::RingBuffer(RingBufferError::NonExistantKey));
        }
```